File: app/src/main/java/com/davismiyashiro/expenses/view/opentab/ExpenseFragment.java
package com.davismiyashiro.expenses.view.opentab;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.davismiyashiro.expenses.R;
import com.davismiyashiro.expenses.datatypes.Expense;
import com.davismiyashiro.expenses.datatypes.Tab;
import com.davismiyashiro.expenses.injection.App;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import timber.log.Timber;

/**
 * A simple {@link Fragment} subclass.
 * Activities that contain this fragment must implement the
 * {@link ExpenseFragment.OnExpenseFragmentInteractionListener} interface
 * to handle interaction events.
 * Use the {@link ExpenseFragment#newInstance} factory method to
 * create an instance of this fragment.
 */
public class ExpenseFragment extends Fragment implements ExpenseInterfaces.ExpenseView{
    private static final String TAB_PARAM = "com.davismiyashiro.expenses.view.opentab.ExpenseFragment";
    private OnExpenseFragmentInteractionListener mListener;

    @Inject
    ExpenseInterfaces.UserActionsListener mPresenter;

    private ExpenseRecyclerViewAdapter mRecyclerAdapter;
    private Tab mTab;
    private RecyclerView recyclerView;
    // TODO: Set Layout of RecyclerView
    private int mColumnCount = 1;

    public ExpenseFragment() {
        // Required empty public constructor
    }

    /**
     * Use this factory method to create a new instance of
     * this fragment using the provided parameters.
     *
     * @param tab Parameter 1.
     * @return A new instance of fragment ExpenseFragment.
     */
    public static ExpenseFragment newInstance(Tab tab) {
        ExpenseFragment fragment = new ExpenseFragment();
        Bundle args = new Bundle();
        args.putParcelable(TAB_PARAM, tab);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Timber.d("onCreate");

        if (getArguments() != null) {
            mTab = getArguments().getParcelable(TAB_PARAM);
        }
        mRecyclerAdapter = new ExpenseRecyclerViewAdapter(new ArrayList<Expense>(), mListener, getActivity());
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        Timber.d("onCreateView");

        View rootView = inflater.inflate(R.layout.fragment_expense, container, false);

        recyclerView = (RecyclerView) rootView.findViewById(R.id.exp_list);

        if (mColumnCount <= 1) {
            recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        } else {
            recyclerView.setLayoutManager(new GridLayoutManager(getContext(), mColumnCount));
        }

//        RecyclerView.ItemDecoration itemDecoration = new
//                DividerItemDecoration(getContext(), DividerItemDecoration.VERTICAL);
//        recyclerView.addItemDecoration(itemDecoration);
//        recyclerView.setItemAnimator(new DefaultItemAnimator());
        recyclerView.setAdapter(mRecyclerAdapter);
        // Inflate the layout for this fragment
        return rootView;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        ((App)getActivity().getApplication()).getComponent().inject(this);
    }

    @Override
    public void onStart() {
        super.onStart();
        mPresenter.setExpenseView(this);
    }

    @Override
    public void onResume() {
        super.onResume();
        Timber.d("onResume");

        mPresenter.loadExpenses(mTab);
    }

    @Override
    public void showExpenses(List<Expense> expenses) {
        mRecyclerAdapter.replaceData(expenses);
    }

    // TODO: Rename method, update argument and hook method into UI event
    public void onButtonPressed(Expense expense) {
        if (mListener != null) {
            mListener.onExpenseFragmentInteraction(expense);
        }
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        Timber.d("onAttach");

        if (context instanceof OnExpenseFragmentInteractionListener) {
            mListener = (OnExpenseFragmentInteractionListener) context;
        } else {
            throw new RuntimeException(context.toString()
                    + " must implement OnExpenseFragmentInteractionListener");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

    /**
     * This interface must be implemented by activities that contain this
     * fragment to allow an interaction in this fragment to be communicated
     * to the activity and potentially other fragments contained in that
     * activity.
     * <p/>
     * See the Android Training lesson <a href=
     * "http://developer.android.com/training/basics/fragments/communicating.html"
     * >Communicating with Other Fragments</a> for more information.
     */
    public interface OnExpenseFragmentInteractionListener {
        void onExpenseFragmentInteraction(Expense expense);
        void onExpenseFragmentLongClick(Expense expense);
    }
}

File: app/src/main/java/com/davismiyashiro/expenses/view/opentab/ExpenseFragment.kt
/*
 * MIT License
 *
 * Copyright (c) 2019 Davis Miyashiro
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package com.davismiyashiro.expenses.view.opentab

import android.app.Activity
import android.content.Context
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.davismiyashiro.expenses.R
import com.davismiyashiro.expenses.datatypes.Expense
import com.davismiyashiro.expenses.datatypes.Tab
import com.davismiyashiro.expenses.injection.App
import com.davismiyashiro.expenses.view.addexpense.ExpenseActivity

import java.util.ArrayList

import javax.inject.Inject

import timber.log.Timber

class ExpenseFragment : Fragment(),
        ExpenseInterfaces.ExpenseView,
        ExpenseRecyclerViewAdapter.OnExpenseFragmentInteractionListener {

    @Inject
    lateinit var mPresenter: ExpenseInterfaces.UserActionsListener

    private lateinit var mRecyclerAdapter: ExpenseRecyclerViewAdapter
    private lateinit var mTab: Tab
    private lateinit var recyclerView: RecyclerView
    // TODO: Set Layout of RecyclerView
    private val mColumnCount = 1

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        Timber.d("onCreate")

        if (arguments != null && arguments is Bundle) {
            mTab = (arguments as Bundle).getParcelable(TAB_PARAM)
        }
        mRecyclerAdapter = ExpenseRecyclerViewAdapter(ArrayList(), this, activity as Activity)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        Timber.d("onCreateView")

        val rootView = inflater.inflate(R.layout.fragment_expense, container, false)

        recyclerView = rootView.findViewById<View>(R.id.exp_list) as RecyclerView

        if (mColumnCount <= 1) {
            recyclerView.layoutManager = LinearLayoutManager(context)
        } else {
            recyclerView.layoutManager = GridLayoutManager(context, mColumnCount)
        }

        recyclerView.adapter = mRecyclerAdapter
        return rootView
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        (activity?.application as App).component.inject(this)
    }

    override fun onStart() {
        super.onStart()
        mPresenter.setExpenseView(this)
    }

    override fun onResume() {
        super.onResume()
        Timber.d("onResume")

        mPresenter.loadExpenses(mTab)
    }

    override fun showExpenses(expenses: MutableList<Expense>) {
        mRecyclerAdapter.replaceData(expenses)
    }

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        Timber.d("onAttach")
    }

    override fun onDetach() {
        super.onDetach()
        Timber.d("onDetach")
    }

    override fun onExpenseFragmentInteraction(expense: Expense) {
        activity?.applicationContext?.let {
            startActivity(ExpenseActivity.newIntent(it, mTab, expense))
        }
    }

    override fun onExpenseFragmentLongClick(expense: Expense) {
        mPresenter.removeExpense(expense)

        mRecyclerAdapter.notifyDataSetChanged()
    }

    companion object {
        private const val TAB_PARAM = "com.davismiyashiro.expenses.view.opentab.ExpenseFragment"

        /**
         * Use this factory method to create a new instance of
         * this fragment using the provided parameters.
         *
         * @param tab Parameter 1.
         * @return A new instance of fragment ExpenseFragment.
         */
        fun newInstance(tab: Tab): ExpenseFragment {
            val fragment = ExpenseFragment()
            val args = Bundle()
            args.putParcelable(TAB_PARAM, tab)
            fragment.arguments = args
            return fragment
        }
    }
} // Required empty public constructor
